src/mame/drivers/alphatpx.cpp

```diff
@@ -585,32 +1015,32 @@ void alphatpx_state::machine_start()
 	m_kbdclk = 0;   // must be initialized here b/c mcs48_reset() causes write of 0xff to all ports
 }
 
-void alphatpx_state::machine_reset()
+void alphatp_12_state::machine_reset()
 {
 	m_kbdread = 1;
 	m_kbdclk = 1;   m_fdc_irq = m_fdc_drq = m_fdc_hld = 0;
 	m_curfloppy = nullptr;
+	// look up floppies in advance
+	m_con1 = machine().device<floppy_connector>("fdc:0");
+	m_con2 = machine().device<floppy_connector>("fdc:1");
+	m_con3 = machine().device<floppy_connector>("fdc:2");
```

Comment:
You really should be able to do this with `optional_device_array` - why are you doing it like this?

Reply:
* The code is copied over from itt3030.cpp where it's done like this.
* optional_device_array is only used in six drivers on the former MAME side of things, so I haven't run into it until now
* after checking the example, I'm lacking the knowledge to see the benefits as 

	for (int i = 0; i < 6; ++i)
		if (m_driver_latch[i].found())
			m_driver_latch[i]->write_bit(offset, BIT(data, i));

(from joctronic.cpp) doesn't seem to save lines.

Reply:
* The `optional_device_array` isn't particularly common, but the conceptually similar `required_device_array` is used in over 30 drivers and a number of devices
* Even if you weren't to use `optional_device_array` you should be using three `optional_device` members instead of naked pointers
* Using device finders is easier to debug when things go wrong because they're handled consistently during the object resolution phase before devices are started
* You have one member declaration `m_floppycon` or whatever instead of three `m_con1`, `m_con2` and `m_con3`
* You have one line in the constructor's initialiser list rather than these three lines in machine start
* You can use loops and indexing on an `optional_device_array` while you need long if-else-if structures which individual members
The example you gave would be 12 lines with individual members like you're using, but it's 3 lines with the `optional_device_array` - a saving of 75%

Reply:
Thanks for that detailed answer! I'll have to look up the examples you cited and see what fits best.